modules/currency-l0/src/main/scala/org/tessellation/currency/l0/snapshot/services/StateChannelBinarySender.scala

```diff
@@ -0,0 +1,238 @@
+package org.tessellation.currency.l0.snapshot.services
+
+import cats.Applicative
+import cats.data.NonEmptyList
+import cats.effect.{Async, Ref}
+import cats.syntax.all._
+
+import scala.collection.immutable.Queue
+
+import org.tessellation.currency.l0.node.IdentifierStorage
+import org.tessellation.node.shared.domain.cluster.storage.L0ClusterStorage
+import org.tessellation.node.shared.domain.snapshot.storage.LastSnapshotStorage
+import org.tessellation.node.shared.domain.statechannel.StateChannelValidator.StateChannelValidationError
+import org.tessellation.node.shared.http.p2p.clients.StateChannelSnapshotClient
+import org.tessellation.schema.{GlobalIncrementalSnapshot, GlobalSnapshotInfo, SnapshotOrdinal}
+import org.tessellation.security.{Hashed, Hasher}
+import org.tessellation.statechannel.StateChannelSnapshotBinary
+
+import eu.timepit.refined.auto._
+import eu.timepit.refined.types.all.NonNegLong
+import eu.timepit.refined.types.numeric.PosLong
+import org.typelevel.log4cats.slf4j.Slf4jLogger
+import retry._
+
+case class TrackedBinary(
+  binary: Hashed[StateChannelSnapshotBinary],
+  enqueuedAtOrdinal: SnapshotOrdinal,
+  sendsSoFar: NonNegLong
+)
+
+case class State(
+  pending: Queue[TrackedBinary],
+  cap: NonNegLong,
+  retryMode: Boolean,
+  noConfirmationsSinceRetryCount: NonNegLong,
+  backoffExponent: NonNegLong
+)
+
+object State {
+  def empty: State = State(
+    pending = Queue.empty[TrackedBinary],
+    cap = 4L,
+    retryMode = false,
+    noConfirmationsSinceRetryCount = 0L,
+    backoffExponent = 0L
+  )
+}
```

Comment:
Not sure if I mentioned this before:

1. These are not part of the interface but only part of the implementation in the companion object, so shouldn't their scope be confined there?
2. Is there any benefit to using refined types for an internal data structure?

Reply:
> Is there any benefit to using refined types for an internal data structure?

being constrained is basically less error-prone and will guide you step by step if you'll come back to the code after a while. imo it's the same reason why you like strictly typing "internal" variables to see the type matching what you expected before writing the implementation. so tl;dr; we should not differentiate whether it's internal or exposed part of the application, refined isn't there for exposing more constrained API, but mostly for our internal usage